lib/Sema/TypeCheckMacros.h

```diff
@@ -17,11 +17,40 @@
 #define SWIFT_SEMA_TYPECHECKMACROS_H
 
 #include "swift/AST/Type.h"
+#include "llvm/ADT/StringRef.h"
 
 namespace swift {
 
 class Expr;
-  
+class TypeRepr;
+
+#if SWIFT_SWIFT_PARSER
+
+namespace macro_context {
+/// Retrieves the evaluation context of a macro with the given name.
+///
+/// The macro evaluation context is a user-defined generic signature and return
+/// type that serves as the "interface type" of references to the macro. The
+/// current implementation takes those pieces of syntax from the macro itself,
+/// then inserts them into a Swift struct that looks like
+///
+/// \code
+/// struct __MacroEvaluationContext\(macro.genericSignature) {
+///   typealias SignatureType = \(macro.signature)
+/// }
+/// \endcode
+///
+/// So that we can use all of Swift's native name lookup and type resolution
+/// facilities to map the parsed signature type back into a semantic \c Type
+/// AST and a set of requiremnets.
+///
+/// \param macroName The name of the macro to look up.
+/// \param useDC The decl context of the use of this macro.
+StructDecl *lookup(StringRef macroName, DeclContext *useDC);
```

Comment:
Not important right now, but I think we should replace the `StructDecl *` with a `MacroDecl` or similar that captures the salient information for the macro and hides the struct-parsing hack a bit.